### DISCUSSION TITLE
Cannot Translate Where Clause on Child Property

### DISCUSSION THREAD

**Discussion by TannerStrike** [AutoMapper/AutoMapper.Extensions.ExpressionMapping#146] [Q&A] [answered]
Hello, I am getting an `System.InvalidOperationException: The LINQ expression could not be translated. Either rewrite the query in a form that can be translated` error.

Here is my setup:

Models:
```csharp
public record CatalogLocation(Guid Id, string Name, CatalogAddress Address);
public record CatalogAddress(string Street1, string PostalCode, string City, string Region, string Country);
public class MerchantLocation {
    public MerchantLocation(Guid id, string name, string street, string city, string region, string country, string postalCode);
}
```

AutoMapper Config:
```csharp
cfg => {
    cfg.AddExpressionMapping();

    cfg.CreateProjection<MerchantLocation, CatalogAddress>()
        .ForCtorParam(nameof(CatalogAddress.Street1), opt => opt.MapFrom(src => src.Street));
    
    cfg.CreateProjection<MerchantLocation, CatalogLocation>()
        .ForCtorParam(nameof(CatalogLocation.Address), opt => opt.MapFrom(src => src));
}
```

The queries in question:
```csharp
// works
    var query1 = _db.MerchantLocations
        .UseAsDataSource(_config)
        .For<CatalogLocation>()
        .Where(cl => cl.Name == "Matthews Avos")
        .ToList();

// cannot be translated
    var query2 = _db.MerchantLocations
        .UseAsDataSource(_config)
        .For<CatalogLocation>()
        .Where(cl => cl.Address.PostalCode == "28226")
	.ToList();
```

When filtering on a top level item, such as `CatalogLocation.Name`, the query works fine. However, when trying to filter on `Address.PostalCode`, it is unable to translate. What should be happening is that `cl.Address.PostalCode` gets mapped to `MerchantLocation.PostalCode`.

I understand that I am filtering on `Address.PostalCode`, which is a product of `.For<CatalogLocation>()` which is not completed until `ToList()` is called. With that being said, I still need this filter, and in my production code I cannot simply call `_db.MerchantLocations.Where(ml => ml.PostalCode == "x")` (the provided example is a contrived example). 

Is this possible? Is there a mapping I can use to make this work? Any help on this would appreciated!

**Comment by BlaiseD** [comment, marked as answer]
Not sure what's going on there -  or the last time anyone maintained that part of the library.

Have you considered using a helper like [this one](https://github.com/AutoMapper/AutoMapper.Extensions.OData/blob/5d83eeb93346c6412cf21c7df768e499ff2bf111/AutoMapper.AspNetCore.OData.EFCore/QueryableExtensions.cs#L152)?

**Reply by TannerStrike** [reply]
Yes, exactly! Meant to put an answer on here, but this ended up being the solution. Instead of trying to trying to apply a set `ODataQueryOptions` of a DTO type to an underlying EF type (which is what I wa truly trying to do), the solution was to convert `ODataQueryOptions`to an expression. Then use AutoMapper to map that expression to the underlying EF type. Then apply the expression to my query, and use AutoMapper again to map the results back to my client side / frontend domain model.

**Reply by BlaiseD** [reply]
Just FYI that link is to [another library](https://www.nuget.org/packages/AutoMapper.AspNetCore.OData.EFCore) which should do that [out of the box](https://github.com/AutoMapper/AutoMapper.Extensions.OData/blob/5d83eeb93346c6412cf21c7df768e499ff2bf111/WebAPI.OData.EFCore/Controllers/OpsTenantController.cs#L46).
